RecoBTag/FeatureTools/src/deep_helpers.cc

```diff
@@ -119,4 +121,108 @@ namespace btagbtvdeep {
     return int16_t((qualityFlags & lostInnerHitsMask) >> lostInnerHitsShift) - 1;
   }
 
+  int center_norm_pad(const std::vector<float> &input,
+                      float center,
+                      float norm_factor,
+                      unsigned min_length,
+                      unsigned max_length,
+                      std::vector<float> &datavec,
+                      int startval,
+                      float pad_value,
+                      float replace_inf_value,
+                      float min,
+                      float max) {
+    // do variable shifting/scaling/padding/clipping in one go
+
+    assert(min <= pad_value && pad_value <= max);
+    assert(min_length <= max_length);
+
+    unsigned target_length = std::clamp((unsigned)input.size(), min_length, max_length);
+    for (unsigned i = 0; i < target_length; ++i) {
+      if (i < input.size()) {
+        datavec[i + startval] = std::clamp((catch_infs(input[i], replace_inf_value) - center) * norm_factor, min, max);
+      } else {
+        datavec[i + startval] = pad_value;
+      }
+    }
+    return target_length;
+  }
+
+  void ParticleNetConstructor(const edm::ParameterSet &Config_,
+			      bool doExtra,
+			      std::vector<std::string> &input_names_,
+			      std::unordered_map<std::string, PreprocessParams> &prep_info_map_,
+			      std::vector<std::vector<int64_t>> &input_shapes_,
+			      std::vector<unsigned> &input_sizes_,
+			      cms::Ort::FloatArrays *data_) {
+    // load preprocessing info
+    auto json_path = Config_.getParameter<std::string>("preprocess_json");
+    if (!json_path.empty()) {
+      // use preprocessing json file if available
+      std::ifstream ifs(edm::FileInPath(json_path).fullPath());
+      nlohmann::json js = nlohmann::json::parse(ifs);
+      js.at("input_names").get_to(input_names_);
+      for (const auto &group_name : input_names_) {
+	const auto &group_pset = js.at(group_name);
```

Comment:
please rerun `scram b code-format` to fix mixture of tabs and spaces for indentation

Reply:
Whoops, yes forgot about that